keps/sig-auth/4412-projected-service-account-tokens-for-kubelet-image-credential-providers/README.md

```diff
@@ -0,0 +1,1034 @@
+   2. The credential provider can be set up multiple times for each audience.
+2. `ServiceAccountAnnotations` is the list of annotations on the service account that plugin can use to extract additional information required to fetch credentials. The
+annotations will be passed to the plugin along with the image and service account token. The plugin is responsible for validating and using the annotations as needed.
+   1. Service account annotations are significantly more stable (and map naturally to the service account being the partioning dimension).
+   2. This field makes the provider opt into getting the annotations it wants, avoids sending arbitrary annotation contents down to the plugin (including stuff like client-side apply annotations) and shrinks the set of annotations that could invalidate the cache.
+
+```diff
+--- a/pkg/kubelet/apis/config/types.go
++++ b/pkg/kubelet/apis/config/types.go
+@@ -653,10 +653,19 @@ type CredentialProvider struct {
+        // +optional
+        Env []ExecEnvVar
+ 
++       // ServiceAccountTokenAudience is the intended audience for the projected service account token.
++       // If set, the kubelet will generate a service account token for the audience and pass it to the plugin.
++       // +optional
++       ServiceAccountTokenAudience string
++
++       // ServiceAccountAnnotations is the list of annotations to be passed to the plugin.
++       // This field is optional and may be empty. Plugins may use this field to extract
++       // additional information required to fetch credentials, for example, identity ID
++       // or identity type to use when fetching credentials.
++       // Only annotations defined in this list will be passed to the plugin and the plugin is 
++       // responsible for validating the annotations.
++       // +optional
++       ServiceAccountAnnotations []string
+```
+
+Example credential provider configuration:
+
+```yaml
+apiVersion: kubelet.config.k8s.io/v1
+kind: CredentialProviderConfig
+providers:
+  - name: acr-credential-provider
+    matchImages:
+      - "*.registry.io/*"
+    defaultCacheDuration: "10m"
+    apiVersion: credentialprovider.kubelet.k8s.io/v1
+    serviceAccountTokenAudience: my-audience
+    serviceAccountAnnotations:
+      - domain.io/identity-id
+      - domain.io/identity-type
+```
+
+##### Kubernetes API Server (KAS) Changes
+
+Add a new flag `--image-pull-credential-provider-audiences` to KAS to allow configuring the list of audiences for which the Kubelet is allowed to generate service account tokens for image pulls.
+
+The node audience restriction will be enforced by the [NodeRestriction admission controller](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#noderestriction)
+and kubelet will only be allowed to generate a service account token for the audience configured in the credential provider configuration if it is present in the list of audiences configured in KAS.
```

Comment:
If this node audience field is only for image pull creds: how will the KAS differentiate between SA tokens the kubelet requests for other purposes?

Reply:
This is the list of allowed audiences the kubelet can request a token for. IIUC, the API server doesn't validate _why_ kubelet is requesting a token for but only checks if the audience is referenced somewhere (in this case configured in the API server as a list of audiences)

xref: https://github.com/kubernetes/kubernetes/compare/master...liggitt:kubernetes:node-audience-restriction?w=1

Reply:
so maybe this flag should be named more generically to reflect that. `--allowed-kubelet-audiences` or something?

Reply:
`--allowed-kubelet-audiences` is fine! I'll update it.